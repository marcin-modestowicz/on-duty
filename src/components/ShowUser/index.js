//@flow
export { default } from "./ShowUser";
