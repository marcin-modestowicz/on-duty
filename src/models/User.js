// @flow
import { AvailabilityCalendar } from "./Calendar";
import { type AvailabilityStatus } from "./Availability";
import uuid from "uuid/v1";

export const POWER = {
  RESIDENT: 0,
  DOCTOR: 1,
  SPECIALIST: 2
};

const REQUIRED_SHIFTS = {
  RESIDENT: 2,
  DOCTOR: 1,
  SPECIALIST: 1
};

class User {
  id: string;
  name: string;
  isDoctor: boolean;
  isSpecialist: boolean;
  isAdmin: boolean;
  availabilityCalendar: AvailabilityCalendar;

  constructor(
    name: string,
    isDoctor?: boolean,
    isSpecialist?: boolean,
    availability?: AvailabilityStatus[]
  ) {
    this.id = uuid();
    this.name = name;
    this.isDoctor = !!isDoctor;
    this.isSpecialist = !!isSpecialist;
    this.isAdmin = false;
    this.availabilityCalendar = new AvailabilityCalendar(availability);
  }

  get power(): number {
    let power = POWER.RESIDENT;

    if (this.isDoctor) {
      power += POWER.DOCTOR;
    }

    if (this.isSpecialist) {
      power += POWER.SPECIALIST;
    }

    return power;
  }

  get requiredShifts(): number {
    let requiredShifts = REQUIRED_SHIFTS.RESIDENT;

    if (this.isDoctor) {
      requiredShifts = REQUIRED_SHIFTS.DOCTOR;
    }

    if (this.isSpecialist) {
      requiredShifts = REQUIRED_SHIFTS.SPECIALIST;
    }

    return requiredShifts;
  }
}

export default User;
