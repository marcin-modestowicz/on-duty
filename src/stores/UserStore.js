//@flow
import { observable, computed, runInAction } from "mobx";
import firebase from "../firebase";
import User from "../models/User";
import { AvailabilityCalendar } from "../models/Calendar";
import { type AvailabilityStatus } from "../models/Availability";

class UserStore {
  @observable user: User;

  constructor(userId: ?string) {
    if (userId) {
      firebase
        .database()
        .ref(`/users/${userId}`)
        .on("value", snapshot => this.getUserData(snapshot));
    }
  }

  @computed
  get isAdmin(): boolean {
    return !!this.user && this.user.isAdmin;
  }

  getUserData(snapshot: Object) {
    const user = snapshot.val();

    if (user) {
      if (!this.user) {
        this.user = new User(
          snapshot.key,
          user.name,
          user.isDoctor,
          user.isSpecialist,
          undefined,
          user.isAdmin
        );
      }

      if (user.availabilityCalendar) {
        this.user.availabilityCalendar.days.forEach(day => {
          const timestamp = day.date.getTime();

          if (
            user.availabilityCalendar[timestamp] != null &&
            user.availabilityCalendar[timestamp] !== day.availability.status
          ) {
            day.availability.setStatus(user.availabilityCalendar[timestamp]);
          }
        });
      }
    }
  }

  setDayStatus = (date: Date, status: AvailabilityStatus) => {
    firebase
      .database()
      .ref(`/users/${this.user.id}/availabilityCalendar/${date.getTime()}`)
      .set(status);
  };

  setAllDaysStatus = (
    calendar: AvailabilityCalendar,
    status: AvailabilityStatus
  ) => {
    const availabilityCalendar = calendar.days.reduce((sum, day) => {
      sum[day.date.getTime()] = status;
      return sum;
    }, {});

    firebase
      .database()
      .ref(`/users/${this.user.id}/availabilityCalendar`)
      .set(availabilityCalendar);
  };
}

export default UserStore;
