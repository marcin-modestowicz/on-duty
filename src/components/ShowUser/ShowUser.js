// @flow
import React, { Component } from "react";
import { observer } from "mobx-react";
import User from "../../models/User";
import UserStore from "../../stores/UserStore";
import AvailabilityCalendar from "../AvailabilityCalendar";

type Props = {
  user: User
};

class ShowUser extends Component<*> {
  props: Props;
  userStore: UserStore;

  constructor(props: Props) {
    super(props);

    this.userStore = new UserStore(props.user);
  }

  render() {
    const user = this.props.user;
    return (
      <div>
        <header>
          <h3>{user.name}</h3>
          <h4>
            {user.isDoctor && "Doctor"} {user.isSpecialist && "Specialist"}
          </h4>
        </header>
        <AvailabilityCalendar
          calendar={user.availabilityCalendar}
          userStore={this.userStore}
        />
      </div>
    );
  }
}

export default ShowUser;
