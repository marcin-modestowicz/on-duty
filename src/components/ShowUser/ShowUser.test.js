// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import User from "../../models/User";
import UserStore from "../../stores/UserStore";
import ShowUser from "./ShowUser";

describe("ShowUser component", () => {
  let RealDate;
  let showUser;

  beforeAll(() => {
    RealDate = Date;

    const date = new Date(2017, 9, 1);
    global.Date = function DateMock(...args) {
      return args.length ? new RealDate(...args) : date;
    };
    const userStore = new UserStore();
    userStore.user = new User("a1", "Marty McFly", true, true);
    showUser = shallow(<ShowUser userStore={userStore} />);
  });

  afterAll(() => {
    global.Date = RealDate;
  });

  test("should render", () => {
    expect(toJson(showUser)).toMatchSnapshot();
  });
});
