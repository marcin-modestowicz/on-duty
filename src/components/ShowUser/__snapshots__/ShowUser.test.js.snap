// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowUser component should render 1`] = `
<div>
  <header>
    <h3>
      Marty McFly
    </h3>
    <h4>
      Doctor
       
      Specialist
    </h4>
  </header>
  <AvailabilityCalendar
    calendar={
      AvailabilityCalendar {
        "days": Array [
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-10-31T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-01T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-02T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-03T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-04T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-05T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-06T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-07T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-08T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-09T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-10T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-11T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-12T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-13T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-14T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-15T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-16T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-17T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-18T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-19T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-20T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-21T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-22T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-23T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-24T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-25T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-26T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-27T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-28T23:00:00.000Z,
          },
          Object {
            "availability": Availability {
              "status": 0,
            },
            "date": 2017-11-29T23:00:00.000Z,
          },
        ],
      }
    }
  />
</div>
`;
